python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,603 @@
+        if max_features is not None:
+            if not isinstance(max_features, int) or max_features <= 0:
+                raise ValueError(
+                    "max_features=%r, neither a positive integer nor None"
+                    % max_features)
+        self.ngram_range = ngram_range
+        self.vocabulary = vocabulary
+        self.binary = binary
+        self.dtype = dtype
+        self.delimiter = delimiter
+        if dtype not in CUPY_SPARSE_DTYPES:
+            msg = f"Expected dtype in {CUPY_SPARSE_DTYPES}, got {dtype}"
+            raise ValueError(msg)
+
+        sklearn_params = {"input": input,
+                          "encoding": encoding,
+                          "decode_error": decode_error,
+                          "strip_accents": strip_accents,
+                          "tokenizer": tokenizer,
+                          "token_pattern": token_pattern}
+        self._check_sklearn_params(analyzer, sklearn_params)
+
+    def _count_vocab(self, tokenized_df):
+        """Count occurrences of tokens in each document."""
+        # Transform string tokens into token indexes from 0 to len(vocab)
+        # The indexes are based on lexicographical ordering.
+        tokenized_df['token'] = tokenized_df['token'].astype('category')
+        tokenized_df['token'] = tokenized_df['token'].cat.set_categories(
+            self.vocabulary_
+        )._column.codes
+
+        # Count of each token in each document
+        count_df = (
+            tokenized_df[["doc_id", "token"]]
+            .groupby(["doc_id", "token"])
+            .size()
+            .reset_index()
+            .rename({0: "count"})
+        )
+
+        return count_df
+
+    def _filter_and_renumber(self, df, keep_values, column):
+        """Filter dataframe to keep only values from column matching
+        keep_values."""
+        df[column] = (
+            df[column].astype('category')
+            .cat.set_categories(keep_values)
+            ._column.codes
+        )
+        df = df.dropna(subset=column)
+        return df
+
+    def _limit_features(self, count_df, vocab, high, low, limit):
+        """Remove too rare or too common features.
+
+        Prune features that are non zero in more samples than high or less
+        documents than low, modifying the vocabulary, and restricting it to
+        at most the limit most frequent.
+        """
+        if high is None and low is None and limit is None:
+            self.stop_words_ = None
+            return count_df
+
+        document_frequency = _document_frequency(count_df)
+
+        mask = cp.ones(len(document_frequency), dtype=bool)
+        if high is not None:
+            mask &= document_frequency <= high
+        if low is not None:
+            mask &= document_frequency >= low
+        if limit is not None and mask.sum() > limit:
+            term_frequency = _term_frequency(count_df)
+            mask_inds = (-term_frequency[mask]).argsort()[:limit]
```

Comment:
How are ties handled? Is argsort deterministic? Would be good to make sure that's well-documented - can be very surprising if there is a bit of nondeterminisim in there

Reply:
Good point, however, from cupy's documentation, argsort use a stable sorting algorithm: https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.argsort.html
Even if it were not a stable sorting (since it might change in the future), it would not swap the values for the same input from one run to an other. Instability doesn't imply randomness, it just means that the input order for values with the same sort key could end up being swapped. When values are swapped is entirely deterministic.